linkerd/app/outbound/src/http/logical.rs

```diff
@@ -1,143 +1,270 @@
                 .push(http::NewTimeout::layer())
                 // Records per-route metrics.
                 .push(
                     rt.metrics
                         .proxy
                         .http_profile_route
-                        .to_layer::<classify::Response, _, _>(),
+                        .to_layer::<classify::Response, _, RouteParams>(),
                 )
                 // Sets the per-route response classifier as a request
                 // extension.
                 .push(classify::NewClassify::layer())
-                .push_on_service(http::BoxResponse::layer())
-                .check_new_service::<ProfileRoute, http::Request<http::BoxBody>>();
-
-            split
-                .push_switch(
-                    |(route, logical): (Option<profiles::http::Route>, Logical)| -> Result<_, Infallible> {
-                        Ok(match route {
-                            None => svc::Either::A(logical),
-                            Some(route) => svc::Either::B(ProfileRoute { route, logical }),
-                        })
-                    },
-                    route.into_inner(),
-                )
-                .push(profiles::http::NewServiceRouter::layer())
-                // Strips headers that may be set by this proxy and add an
-                // outbound canonical-dst-header. The response body is boxed
-                // unify the profile stack's response type with that of to
-                // endpoint stack.
+                .push_on_service(http::BoxResponse::layer());
+
+            // A `NewService`--instantiated once per logical target--that caches
+            // a set of concrete services so that, as the watch provides new
+            // `Params`, we can reuse inner services.
+            let router = svc::layers()
+                // Each `RouteParams` provides a `Distribution` that is used to
+                // choose a concrete service for a given route.
+                .push(BackendCache::layer())
+                // Lazily cache a service for each `RouteParams`
+                // returned from the `SelectRoute` impl.
+                .push_on_service(route)
+                .push(svc::NewOneshotRoute::<Params, _, _>::layer_cached());
+
+            // For each `Logical` target, watch its `Profile`, rebuilding a
+            // router stack.
+            concrete
+                // Share the concrete stack with each router stack.
+                .push_new_clone()
+                // Rebuild this router stack every time the profile changes.
+                .push_on_service(router)
+                .push(svc::NewSpawnWatch::<Profile, _>::layer_into::<Params>())
+                // Add l5d-dst-canonical header to requests.
+                //
+                // TODO(ver) move this into the endpoint stack so that we can only
+                // set this on meshed connections.
+                //
+                // TODO(ver) do we need to strip headers here?
```

Comment:
do we need to figure this out before merging? the previous comment implied that we did this but i'm not sure if that was actually true...

Reply:
Right. The previous comment is wrong, I think. Needs revisiting before stable but not a blocker for now.